qiskit/circuit/quantumcircuit.py

```diff
@@ -1451,55 +1409,30 @@ def qasm(self, formatted=False, filename=None, encoding=None):
                     bit_labels[qubit],
                     bit_labels[clbit],
                 )
-
-            # If instruction is a root gate or a root instruction (in that case, compositive)
-
-            elif (
-                type(instruction)
-                in [
-                    Gate,
-                    Instruction,
-                ]
-                or (isinstance(instruction, ControlledGate) and instruction._open_ctrl)
-            ):
-                if instruction not in existing_composite_circuits:
-                    if instruction.name in existing_gate_names:
-                        old_name = instruction.name
-                        instruction.name += "_" + str(id(instruction))
-
-                        warnings.warn(
-                            "A gate named {} already exists. "
-                            "We have renamed "
-                            "your gate to {}".format(old_name, instruction.name)
-                        )
-
-                    # Get qasm of composite circuit
-                    qasm_string = self._get_composite_circuit_qasm_from_instruction(instruction)
-
-                    # Insert composite circuit qasm definition right after header and extension lib
-                    string_temp = string_temp.replace(
-                        self.extension_lib, f"{self.extension_lib}\n{qasm_string}"
-                    )
+            else:
+                # decompose gate using definitions if they are not defined in OpenQASM2
+                if (
+                    instruction.name not in existing_gate_names
+                    and instruction not in existing_composite_circuits
+                ):
+                    if instruction.name in [
+                        instruction.name for instruction in existing_composite_circuits
+                    ]:
+                        # append instruction id to name to make it unique
+                        instruction.name += f"_{id(instruction)}"
```

Comment:
This seems to modify the circuit data, is that correct? If yes, can we avoid that?

Reply:
Indeed circuit data is modified here. I think the problem is we are using name to identify gates / instructions. We can't distinguish gates if they have the same name but different definitions. I can revert this and leave the handling of gates with same name out of this PR. What do you think?

Reply:
We need the instructions in `existing_composite_circuits` to have unique names right? So we could just store a copy of the instruction where we modified the name instead of storing the same instance that lives in circuit.data 🙂

Reply:
Yes, you are right. Let me fix this.

Reply:
I see that this is the same as the original code. Can I leave this part out of this PR?

Reply:
Yes 👍🏻 could you open an issue for it though?

Reply:
Sure, I'll do that.

Reply:
Issue created here: https://github.com/Qiskit/qiskit-terra/issues/6687 Please add comments if you think the description is not accurate.

Reply:
Perfect, thanks a lot!